453-skeleton/Texture.cpp

```diff
@@ -0,0 +1,88 @@
+#include "Texture.h"
+#define STB_IMAGE_IMPLEMENTATION
+#include <stb/stb_image.h>
+
+#include <iostream>
+
+Texture::Texture(std::string path, GLint interpolation)
+	: textureID(), path(path), interpolation(interpolation)
+{
+	int numComponents;
+	stbi_set_flip_vertically_on_load(true);
+	const char* pathData = path.c_str();
+	unsigned char* data = stbi_load(pathData, &width, &height, &numComponents, 0);
+	if (data != nullptr)
+	{
+		glPixelStorei(GL_UNPACK_ALIGNMENT, 1);		//Set alignment to be 1
+
+		textureID.bind();
+
+		//Set number of components by format of the texture
+		GLuint format = GL_RGB;
+		switch (numComponents)
+		{
+		case 4:
+			format = GL_RGBA;
+			break;
+		case 3:
+			format = GL_RGB;
+			break;
+		case 2:
+			format = GL_RG;
+			break;
+		case 1:
+			format = GL_RED;
+			break;
+		default:
+			std::cout << "Invalid Texture Format" << std::endl;
+			break;
+		};
+		//Loads texture data into bound texture
+		glTexImage2D(GL_TEXTURE_2D, 0, format, width, height, 0, format, GL_UNSIGNED_BYTE, data);
+
+		//Modifies behaviour for bound texture
+		// Note: Only wrapping modes supported for GL_TEXTURE_RECTANGLE when defining
+		// GL_TEXTURE_WRAP are GL_CLAMP_TO_EDGE or GL_CLAMP_TO_BORDER
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, interpolation);
+		glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, interpolation);
+
+		// Clean up
+		textureID.unbind();
+		glPixelStorei(GL_UNPACK_ALIGNMENT, 4);	//Return to default alignment
+		stbi_image_free(data);
+
+		if (CheckGLErrors((std::string("Loading texture: ") + path).c_str())) {
+			throw std::runtime_error("Texture failed to load!");
+		}
+	}
+	else {
+		throw std::runtime_error("Failed to read texture data from file!");
+	}
+}
+
+bool Texture::CheckGLErrors(const char* errorLocation)
```

Comment:
We already have the GLDebug output, I would assume that catches all of these errors already. I don't mind leaving this here if you have any concerns that it won't catch it.

Reply:
Nah, I just copied that function in from the boilerplate from when I took the course. I'll remove it.